src/vcpkg/base/messages.cpp

```diff
@@ -95,17 +96,160 @@ namespace vcpkg
 
     LocalizedString::LocalizedString(StringView data) : m_data(data.data(), data.size()) { }
     LocalizedString::LocalizedString(std::string&& data) noexcept : m_data(std::move(data)) { }
+
 }
 
 namespace vcpkg::msg
 {
-    REGISTER_MESSAGE(SeeURL);
-    REGISTER_MESSAGE(NoteMessage);
-    REGISTER_MESSAGE(WarningMessage);
-    REGISTER_MESSAGE(ErrorMessage);
-    REGISTER_MESSAGE(InternalErrorMessage);
-    REGISTER_MESSAGE(InternalErrorMessageContact);
+    namespace
+    {
+        template<class T>
+        struct ArgExample;
+    }
+}
+#define DECLARE_MSG_ARG(NAME, EXAMPLE)                                                                                 \
+    StringLiteral vcpkg::msg::NAME##_t::name = #NAME;                                                                  \
+    template<>                                                                                                         \
+    struct vcpkg::msg::ArgExample<::vcpkg::msg::NAME##_t>                                                              \
+    {                                                                                                                  \
+        static constexpr StringLiteral example = sizeof(EXAMPLE) > 1 ? StringLiteral("{" #NAME "} is " EXAMPLE ".")    \
+                                                                     : StringLiteral("");                              \
+    };
+#include <vcpkg/base/message-args.inc.h>
+#undef DECLARE_MSG_ARG
+
+namespace vcpkg
+{
+    namespace
+    {
+        static constexpr const size_t max_number_of_args = 5;
+
+        struct MessageData
+        {
+            StringLiteral name;
+            std::array<const StringLiteral*, max_number_of_args> arg_examples;
```

Comment:
StringLiteral is already Pointer-y, should this be `StringLiteral` instead?

Reply:
`StringLiteral` is two pointers instead of one, and since this is _ultra cold_ data I wanted to minimize it.